torch/nn/parallel/data_parallel.py

```diff
@@ -1,10 +1,67 @@
+import weakref
 import torch
+import torch.cuda.comm as comm
+from collections import OrderedDict
 from ..modules import Module
 from .scatter_gather import scatter_kwargs, gather
 from .replicate import replicate
 from .parallel_apply import parallel_apply
 
 
+# NOTE: these callbacks are one-off
+# NOTE: they are dedupliacted based on identity
+class _CallbackOrderedDict(OrderedDict):
+    @classmethod
+    def _init_class(cls):
+        mutating_methods = [
+            '__setitem__',
+            '__delitem__',
+            'pop',
+            'popitem',
+            'update',
+            'setdefault'
+        ]
+
+        def method_wrapper(name):
+            def callback_method(self, *args, **kwargs):
+                for callback in self.callbacks:
+                    callback()
+                del self.callbacks[:]
+                return getattr(super(_CallbackOrderedDict, self), name)(*args, **kwargs)
+            return callback_method
+        for name in mutating_methods:
+            setattr(cls, name, method_wrapper(name))
+
+    @classmethod
+    def from_ordered(cls, od):
+        return cls(od.items())
+
+    def __init__(self, *args, **kwargs):
+        self.callbacks = []
+        super(_CallbackOrderedDict, self).__init__(*args, **kwargs)
+
+    def register_modification_callback(self, cb):
+        if all(c is not cb for c in self.callbacks):
+            self.callbacks.append(cb)
+
+    # Don't serialize callbacks. If a module registers them on its submodules,
+    # it will have to redo it after reloading.
+    def __reduce__(self):
+        state = super(_CallbackOrderedDict, self).__reduce__()
+        # NB: 2nd element of reduce tuple is the dict.
+        if state[2]:
+            lstate = list(state)
+            lstate[2] = state[2].copy().pop('callbacks')
```

Comment:
I'm a little confused (my ignorance here). Doesn't this seem to imply that the pickle *does* return the callbacks?

Reply:
I don't understand the question. Third element of this tuple is just the `__dict__` of this instance. We don't want to mutate it directly, so we copy it, remove callbacks, and serialize only this.